main/lsp/requests/document_formatting.cc

```diff
@@ -30,35 +59,48 @@ void DocumentFormattingTask::index(LSPIndexer &index) {
-                // Non-fatal error. Returns null for result.
-                config.logger->debug("Rubyfmt returned non-fatal error code `{}` for file `{}`", formatResult.status,
-                                     params->textDocument->uri);
-                break;
-            case experimental::rubyfmt::RubyfmtFormatError::RUBYFMT_FORMAT_ERROR_OK:
+        vector<unique_ptr<TextEdit>> edits;
+
+        auto process = subprocess::Popen({"rubyfmt", std::string(index.getFile(fref).path()).c_str()},
+                                         subprocess::output{subprocess::PIPE}, subprocess::input{subprocess::PIPE});
+        process.start_process();
+        auto processResponse = process.communicate();
+        process.wait();
+        auto returnCode = process.retcode();
+
+        std::string formattedContents(processResponse.first.buf.begin(), processResponse.first.buf.end());
+
+        switch (returnCode) {
+            case RubyfmtStatus::OK:
                 // Construct text edit to replace entire document.
-                vector<unique_ptr<TextEdit>> edits;
-                // Note: VS Code uses 0-indexed lines, so the lineCount will be one more line than the size of the doc.
+                // Note: VS Code uses 0-indexed lines, so the lineCount will be one more line than the size of the
+                // doc.
                 edits.emplace_back(
                     make_unique<TextEdit>(make_unique<Range>(make_unique<Position>(0, 0),
                                                              make_unique<Position>(index.getFile(fref).lineCount(), 0)),
-                                          formatResult.formatted));
+                                          formattedContents));
                 result = move(edits);
                 break;
+            case RubyfmtStatus::SYNTAX_ERROR:
+            case RubyfmtStatus::RIPPER_PARSE_FAILURE:
+                // Non-fatal error. Returns null for result.
+                config.logger->debug("Rubyfmt returned non-fatal error code `{}` for file `{}`", process.retcode(),
+                                     params->textDocument->uri);
+                break;
+            case RubyfmtStatus::IO_ERROR:
+                // Fatal error -- crash
+                Exception::raise("Rubyfmt reported an IO error");
+            case RubyfmtStatus::OTHER_RUBY_ERROR:
+                // Fatal error -- crash
+                Exception::raise("Rubyfmt reported a Ruby error");
+            case RubyfmtStatus::INITIALIZE_ERROR:
+                // Fatal error -- crash
+                Exception::raise("Rubyfmt failed to initialize");
+            case RubyfmtStatus::RUBYFMT_NOT_IN_PATH:
+                Exception::raise("`rubyfmt` not found in the PATH.");
+            default:
+                Exception::raise("An unknown exception (exit code {}) occurred: {}", returnCode,
+                                 std::string(processResponse.second.buf.begin(), processResponse.second.buf.end()));
```

Comment:
This will crash the whole LSP server, which seems a big aggressive.

It might be better to return an LSP RPC error. You can grep for `->error =` in `main/lsp/requests/` to see some examples of where we do that. (There's even an example earlier up in this file.)

There's a list of pre-declared error codes we can use here:

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#errorCodes

LSP RPC errors are usually shown

Reply:
I see that you're just cargo-culting what existed before. I'm happy to defer this for the current PR, but it's something that i would like to see fixed before we take this out of beta.